src/main/groovy/liquibase/sdk/test/change/ChangeObjectTests.groovy

```diff
@@ -0,0 +1,153 @@
+
+        if (expectedSql != null && !testInput.pathToChangeLogFile.endsWith(".sql")) {
+            assert generatedSql == expectedSql: "Expected SQL does not match actual sql. Deleting the existing expectedSql file will test that the new SQL works correctly and will auto-generate a new version if it passes"
+            if (!TestConfig.instance.revalidateSql) {
+                return //sql is right. Nothing more to test
+            }
+        }
+
+        assert testInput.database.getConnection() instanceof JdbcConnection: "We cannot verify the following SQL works works because the database is offline:\n${generatedSql}"
+
+        when:
+        liquibase.update(testInput.context)
+
+        String jsonSnapshot = SnapshotHelpers.getJsonSnapshot(testInput.database, catalogAndSchemaList)
+        liquibase.rollback(liquibase.databaseChangeLog.changeSets.size(), testInput.context)
+
+        then:
+        snapshotMatchesSpecifiedStructure(expectedSnapshot, jsonSnapshot)
+
+        if (expectedSql == null && !testInput.pathToChangeLogFile.endsWith(".sql")) {
+            //save generated sql as expected sql for future runs
+            saveAsExpectedSql(generatedSql, testInput)
+        }
+
+        where:
+        testInput << buildTestInput()
+    }
+
+    /**
+     * Standardizes sql content. Removes line ending differences, and unnessisary leading/trailing whitespace
+     * @param sql
+     * @return
+     */
+    protected String cleanSql(String sql) {
+        if (sql == null) {
+            return null
+        }
+        return StringUtil.trimToNull(sql.replace("\r", "")
+                .replaceAll(/(?m)^--.*/, "") //remove comments
+                .replaceAll(/(?m)^\s+/, "") //remove beginning whitepace per line
+                .replaceAll(/(?m)\s+$/, "") //remove trailing whitespace per line
+        ) //remove trailing whitepace per line
+    }
+
+    static void snapshotMatchesSpecifiedStructure(String expected, String actual) {
+        JSONAssert.assertEquals(expected, actual, new SnapshotHelpers.GeneralSnapshotComparator())
+    }
+
+    void saveAsExpectedSql(String generatedSql, TestInput testInput) {
+        File outputFile = "${outputResourcesBase}/liquibase/sdk/test/change/expectedSql/${testInput.database.shortName}/${testInput.changeObject}.sql" as File
+        outputFile.parentFile.mkdirs()
+
+        outputFile.write(generatedSql)
+    }
+
+    List<TestInput> buildTestInput() {
+        //TODO: Handle changeObject selection
+        String inputFormat = System.getProperty("inputFormat")
+        String changeObjects = System.getProperty("changeObjects")
+        if (inputFormat && (!supportedChangeLogFormats.contains(inputFormat))) {
+            throw new IllegalArgumentException(inputFormat + " inputFormat is not supported")
+        }
+
+        Logger.getLogger(this.class.name).warning("Only " + inputFormat + " input files are taken into account for this test run")
+
+//        if (changeObjects) {
```

Comment:
this breaks functionality of providing changelogs as cmd argument (-DchangeObjects=addPrimaryKey)